profiler/src/ProfilerEngine/Datadog.Profiler.Native.Windows/EtwEventsManager.cpp

```diff
@@ -0,0 +1,396 @@
+        }
+
+        auto pThreadInfo = GetOrCreate(tid);
+        if (pThreadInfo->LastEventWasContentionStart())
+        {
+            AttachContentionCallstack(pThreadInfo, cbEventData, pEventData);
+        }
+        else if (pThreadInfo->LastEventWasAllocationTick())
+        {
+            AttachAllocationCallstack(pThreadInfo, cbEventData, pEventData);
+        }
+
+        pThreadInfo->ClearLastEventId();
+    }
+    else if (keyword == KEYWORD_CONTENTION)
+    {
+        if (_pContentionListener == nullptr)
+        {
+            return;
+        }
+
+        if (id == EVENT_CONTENTION_START)
+        {
+            if (_isDebugLogEnabled)
+            {
+                std::cout << "ContentionStart" << std::endl;
+            }
+
+            auto pThreadInfo = GetOrCreate(tid);
+            pThreadInfo->ContentionStartTimestamp = systemTimestamp;
+            pThreadInfo->LastEventId = EventId::ContentionStart;
+        }
+        else if (id == EVENT_CONTENTION_STOP)
+        {
+            if (_isDebugLogEnabled)
+            {
+                std::cout << "ContentionStop" << std::endl;
+            }
+
+            auto pThreadInfo = Find(tid);
+            if (pThreadInfo != nullptr)
+            {
+                pThreadInfo->ClearLastEventId();
+                if (pThreadInfo->ContentionStartTimestamp != 0)
+                {
+                    auto timestamp = TimestampToEpochNS(systemTimestamp); // systemTimestamp is in 100ns units
+                    auto duration = (systemTimestamp - pThreadInfo->ContentionStartTimestamp) * 100;
+                    pThreadInfo->ContentionStartTimestamp = 0;
+
+                    _pContentionListener->OnContention(timestamp, tid, duration, pThreadInfo->ContentionCallStack);
+                }
+            }
+            else
+            {
+                // this should never happen
+                if (_isDebugLogEnabled)
+                {
+                    std::cout << "Missing thread info..." << std::endl;
+                }
+            }
+        }
+    }
+    else if (keyword == KEYWORD_GC)
+    {
+
+        if (id == EVENT_ALLOCATION_TICK)
+        {
+            if (_isDebugLogEnabled)
+            {
+                //std::cout << "AllocationTick" << std::endl;
+            }
+
+            if (_pAllocationListener == nullptr)
+            {
+                return;
+            }
+
+            auto pThreadInfo = GetOrCreate(tid);
+            pThreadInfo->LastEventId = EventId::AllocationTick;
+
+            auto timestamp = TimestampToEpochNS(systemTimestamp);
+            // TODO: get the type name and ClassID from the payload
+            // TODO: update IAllocationListener to take the type name, ClassID and stack
+        }
+        else
+        {
+            // reuse GC events parser
+            auto timestamp = TimestampToEpochNS(systemTimestamp);
+            _parser.get()->ParseEvent(timestamp, version, keyword, id, cbEventData, pEventData);
```

Comment:
I wonder if `_parser->ParseEvent(...)` is not enough ?

Reply:
✅ done